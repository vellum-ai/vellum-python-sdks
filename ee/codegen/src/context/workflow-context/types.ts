export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
