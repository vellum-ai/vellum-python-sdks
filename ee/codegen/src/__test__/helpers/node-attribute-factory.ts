import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
