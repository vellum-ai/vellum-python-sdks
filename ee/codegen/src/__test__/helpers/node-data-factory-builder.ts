import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
