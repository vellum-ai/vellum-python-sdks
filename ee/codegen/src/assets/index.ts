// This directory contains generated assets that are copied during postinstall
// and included in the npm package build. The node-definitions.json file is
// generated from the Python SDK and copied here for use in code generation.
