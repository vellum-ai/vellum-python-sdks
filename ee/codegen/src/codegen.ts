import {
  ErrorLogFile,
  InitFile,
  Inputs,
  NodeInput,
  VellumValue,
  VellumVariable,
  Workflow,
} from "./generators";
import { BasePersistedFile } from "./generators/base-persisted-file";
import { State } from "./generators/state";
import { WorkflowSandboxFile } from "./generators/workflow-sandbox-file";

export function vellumVariable(args: VellumVariable.Args): VellumVariable {
  return new VellumVariable(args);
}

export function vellumValue(args: VellumValue.Args): VellumValue {
  return new VellumValue(args);
}

export function nodeInput(args: NodeInput.Args): NodeInput {
  return new NodeInput(args);
}

export function inputs(args: Inputs.Args): Inputs {
  return new Inputs(args);
}

export function state(args: State.Args): State {
  return new State(args);
}

export function workflow(args: Workflow.Args): Workflow {
  return new Workflow(args);
}

export function initFile(args: InitFile.Args): InitFile {
  return new InitFile(args);
}

export function errorLogFile(args: BasePersistedFile.Args): ErrorLogFile {
  return new ErrorLogFile(args);
}

export function workflowSandboxFile(
  args: WorkflowSandboxFile.Args
): WorkflowSandboxFile {
  return new WorkflowSandboxFile(args);
}
