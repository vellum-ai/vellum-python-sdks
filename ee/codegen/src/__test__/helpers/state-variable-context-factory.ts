import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
