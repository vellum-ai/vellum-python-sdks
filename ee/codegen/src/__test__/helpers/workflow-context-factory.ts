import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
