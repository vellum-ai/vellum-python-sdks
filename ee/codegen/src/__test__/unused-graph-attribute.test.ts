import { Writer } from "@fern-api/python-ast/core/Writer";

import { workflowContextFactory } from "./helpers";
import {
  EdgeFactoryNodePair,
  edgesFactory,
} from "./helpers/edge-data-factories";
import {
  entrypointNodeDataFactory,
  genericNodeFactory,
} from "./helpers/node-data-factories";

import * as codegen from "src/codegen";
import { createNodeContext } from "src/context";
import { WorkflowDataNode } from "src/types/vellum";

describe("Workflow", () => {
  const entrypointNode = entrypointNodeDataFactory();
  describe("unused_graphs", () => {
    const runUnusedGraphsWorkflowTest = async (
      edges: EdgeFactoryNodePair[]
    ) => {
      const workflowContext = workflowContextFactory();
      const writer = new Writer();
      workflowContext.addEntrypointNode(entrypointNode);

      const nodes = Array.from(
        new Set(
          edges
            .flatMap(([source, target]) => [
              Array.isArray(source) ? source[0] : source,
              Array.isArray(target) ? target[0] : target,
            ])
            .filter(
              (node): node is WorkflowDataNode => node.type !== "ENTRYPOINT"
            )
        )
      );

      await Promise.all(
        nodes.map((node) => {
          createNodeContext({
            workflowContext,
            nodeData: node,
          });
        })
      );

      workflowContext.addWorkflowEdges(edgesFactory(edges));

      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName: "test",
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    };

    it("should be empty when all nodes are connected to entrypoint", async () => {
      const connectedNode1 = genericNodeFactory({
        label: "ConnectedNode1",
      });
      const connectedNode2 = genericNodeFactory({
        label: "ConnectedNode2",
      });

      await runUnusedGraphsWorkflowTest([
        [entrypointNode, connectedNode1],
        [connectedNode1, connectedNode2],
      ]);
    });

    it("should identify simple disconnected graph", async () => {
      const connectedNode = genericNodeFactory({
        label: "ConnectedNode",
      });

      const disconnectedNode1 = genericNodeFactory({
        label: "DisconnectedNode1",
      });

      const disconnectedNode2 = genericNodeFactory({
        label: "DisconnectedNode2",
      });

      await runUnusedGraphsWorkflowTest([
        [entrypointNode, connectedNode],
        [disconnectedNode1, disconnectedNode2],
      ]);
    });

    it("should identify multiple disconnected graphs", async () => {
      const connectedNode = genericNodeFactory({
        label: "ConnectedNode",
      });

      const disconnectedNode1 = genericNodeFactory({
        label: "DisconnectedNode1",
      });

      const disconnectedNode2 = genericNodeFactory({
        label: "DisconnectedNode2",
      });

      const disconnectedNode3 = genericNodeFactory({
        label: "DisconnectedNode3",
      });

      const disconnectedNode4 = genericNodeFactory({
        label: "DisconnectedNode4",
      });

      const disconnectedNode5 = genericNodeFactory({
        label: "DisconnectedNode5",
      });

      const disconnectedNode6 = genericNodeFactory({
        label: "DisconnectedNode6",
      });

      await runUnusedGraphsWorkflowTest([
        [entrypointNode, connectedNode],
        [disconnectedNode1, disconnectedNode2],
        [disconnectedNode1, disconnectedNode3],
        [disconnectedNode4, disconnectedNode5],
        [disconnectedNode5, disconnectedNode6],
      ]);
    });

    it("should identify multiple disconnected graphs with flipped edges", async () => {
      const connectedNode = genericNodeFactory({
        label: "ConnectedNode",
      });

      const disconnectedNode1 = genericNodeFactory({
        label: "DisconnectedNode1",
      });

      const disconnectedNode2 = genericNodeFactory({
        label: "DisconnectedNode2",
      });

      const disconnectedNode3 = genericNodeFactory({
        label: "DisconnectedNode3",
      });

      await runUnusedGraphsWorkflowTest([
        [entrypointNode, connectedNode],
        [disconnectedNode2, disconnectedNode3],
        [disconnectedNode1, disconnectedNode2],
      ]);
    });

    it("should handle circular graphs", async () => {
      const connectedNode = genericNodeFactory({
        label: "ConnectedNode",
      });

      const disconnectedNode1 = genericNodeFactory({
        label: "DisconnectedNode1",
      });

      const disconnectedNode2 = genericNodeFactory({
        label: "DisconnectedNode2",
      });

      const disconnectedNode3 = genericNodeFactory({
        label: "DisconnectedNode3",
      });

      await runUnusedGraphsWorkflowTest([
        [entrypointNode, connectedNode],
        [disconnectedNode1, disconnectedNode2],
        [disconnectedNode2, disconnectedNode3],
        [disconnectedNode3, disconnectedNode1],
      ]);
    });
  });
});
