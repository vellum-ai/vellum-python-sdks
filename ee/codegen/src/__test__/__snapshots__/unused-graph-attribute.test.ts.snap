// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Workflow > unused_graphs > should be empty when all nodes are connected to entrypoint 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.my_custom_node import ConnectedNode1
from .nodes.my_custom_node_1 import ConnectedNode2


class TestWorkflow(BaseWorkflow):
    graph = ConnectedNode1 >> ConnectedNode2
"
`;

exports[`Workflow > unused_graphs > should handle circular graphs 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.my_custom_node import ConnectedNode
from .nodes.my_custom_node_1 import DisconnectedNode1
from .nodes.my_custom_node_2 import DisconnectedNode2
from .nodes.my_custom_node_3 import DisconnectedNode3


class TestWorkflow(BaseWorkflow):
    graph = ConnectedNode
    unused_graphs = {
        DisconnectedNode1 >> DisconnectedNode2 >> DisconnectedNode3 >> DisconnectedNode1
    }
"
`;

exports[`Workflow > unused_graphs > should handle conditional node with prompt nodes in different ports in unused graphs 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.conditional_node import ConditionalNode
from .nodes.my_custom_node import ConnectedNode
from .nodes.prompt_node import PromptNode
from .nodes.prompt_node_1 import PromptNode1


class TestWorkflow(BaseWorkflow):
    graph = ConnectedNode
    unused_graphs = {
        ConditionalNode.Ports.if_1 >> PromptNode,
        ConditionalNode.Ports.else_1 >> PromptNode1,
    }
"
`;

exports[`Workflow > unused_graphs > should identify multiple disconnected graphs 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.my_custom_node import ConnectedNode
from .nodes.my_custom_node_1 import DisconnectedNode1
from .nodes.my_custom_node_2 import DisconnectedNode2
from .nodes.my_custom_node_3 import DisconnectedNode3
from .nodes.my_custom_node_4 import DisconnectedNode4
from .nodes.my_custom_node_5 import DisconnectedNode5
from .nodes.my_custom_node_6 import DisconnectedNode6


class TestWorkflow(BaseWorkflow):
    graph = ConnectedNode
    unused_graphs = {
        DisconnectedNode1
        >> {
            DisconnectedNode2,
            DisconnectedNode3,
        },
        DisconnectedNode4 >> DisconnectedNode5 >> DisconnectedNode6,
    }
"
`;

exports[`Workflow > unused_graphs > should identify multiple disconnected graphs with flipped edges 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.my_custom_node import ConnectedNode
from .nodes.my_custom_node_1 import DisconnectedNode2
from .nodes.my_custom_node_2 import DisconnectedNode3
from .nodes.my_custom_node_3 import DisconnectedNode1


class TestWorkflow(BaseWorkflow):
    graph = ConnectedNode
    unused_graphs = {DisconnectedNode1 >> DisconnectedNode2 >> DisconnectedNode3}
"
`;

exports[`Workflow > unused_graphs > should identify simple disconnected graph 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.my_custom_node import ConnectedNode
from .nodes.my_custom_node_1 import DisconnectedNode1
from .nodes.my_custom_node_2 import DisconnectedNode2


class TestWorkflow(BaseWorkflow):
    graph = ConnectedNode
    unused_graphs = {DisconnectedNode1 >> DisconnectedNode2}
"
`;
